polar-core/src/data_filtering.rs

```diff
@@ -1158,3 +1223,447 @@ mod test {
+        match t.value() {
+            Value::Expression(Operation {
+                operator: Dot,
+                args,
+            }) if args.len() == 2 && args[1].value().as_string().is_ok() => {
+                let mut p = Self::path_disasm(&args[0])?;
+                p.push(args[1].value().as_string().unwrap().to_string());
+                Ok(p)
+            }
+            Value::Variable(Symbol(s)) => Ok(vec![s.clone()]),
+            Value::ExternalInstance(ExternalInstance {
+                repr: Some(repr), ..
+            }) => {
+                eprintln!("{}", repr);
+                // FIXME hack for Ruby, obviously wrong
+                let re = regex::Regex::new(r"^#<([^:]+):.*").unwrap();
+                match re
+                    .captures(repr)
+                    .and_then(|cap| cap.get(1))
+                    .map(|c| c.as_str().to_owned())
+                {
+                    Some(tag) => Ok(vec![tag]),
+                    None => err_unimplemented(format!("QueryInfo::path_disasm({})", t.to_polar())),
+                }
+            }
+            _ => err_unimplemented(format!("QueryInfo::path_disasm({})", t.to_polar())),
+        }
+    }
+
+    fn data_source(&self, mut path: Vec<String>) -> DataSource {
+        let base = path.remove(0);
+        let base = self.entities.get(&vec![base]).unwrap().clone();
+        DataSource { base, path }
+    }
+
+    fn datum(&mut self, t: &Term) -> PolarResult<PreDatum> {
+        use {Operator::*, PreDatum::*};
+        match t.value() {
+            Value::Number(_) | Value::String(_) => Ok(Imm(t.value().clone())),
+            Value::Expression(Operation {
+                operator: Dot,
+                args,
+            }) if args.len() == 2 && args[1].value().as_string().is_ok() => Ok(Field(
+                Proj(
+                    self.source(&args[0])?,
+                    args[1].value().as_string()?.to_string(),
+                ),
+                self.data_source(Self::path_disasm(&args[0])?),
+            )),
+            _ => err_unimplemented(format!("datum({})", t.to_polar())),
+        }
+    }
+
+    fn binary_op(mut self, args: &[Term], op: SelOp) -> PolarResult<Self> {
+        let (l, r) = (self.datum(&args[0])?, self.datum(&args[1])?);
+        self.constraints.push((l, op, r));
+        Ok(self)
+    }
+
+    fn source(&mut self, t: &Term) -> PolarResult<Rc<DataFilter>> {
+        use DataFilter::Source;
+        let path = Self::path_disasm(t)?;
+        match self.entities.get(&path) {
+            Some(src) => Ok(Rc::new(Source(src.clone()))),
+            _ => err_invalid(format!("No entity for {:?}", path)),
+        }
+    }
+
+    fn new(types: Types, parts: Vec<Operation>) -> PolarResult<Self> {
+        use Operator::*;
+        let (isas, othas): (Vec<_>, Vec<_>) = parts
+            .into_iter()
+            .partition(|x| matches!(x, Operation { operator: Isa, args } if args.len() == 2));
+        let mut entities = isas
```

Comment:
Is there always a guarantee that there will be an Isa for every subfield that is filtered on in the partial?